File: frontend/appflowy_flutter/lib/plugins/database_view/application/field/field_service.dart
import 'dart:typed_data';

import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/protobuf.dart';
import 'package:appflowy_backend/protobuf/flowy-error/errors.pb.dart';
import 'package:dartz/dartz.dart';

/// FieldService consists of lots of event functions. We define the events in the backend(Rust),
/// you can find the corresponding event implementation in event_map.rs of the corresponding crate.
///
/// You could check out the rust-lib/flowy-database/event_map.rs for more information.
class FieldBackendService {
  FieldBackendService({required this.viewId, required this.fieldId});

  final String viewId;
  final String fieldId;

  static Future<Either<TypeOptionPB, FlowyError>> createField({
    required String viewId,
    FieldType fieldType = FieldType.RichText,
    String? fieldName,
    Uint8List? typeOptionData,
    OrderObjectPositionPB? position,
  }) {
    final payload = CreateFieldPayloadPB(
      viewId: viewId,
      fieldType: fieldType,
      fieldName: fieldName,
      typeOptionData: typeOptionData,
      fieldPosition: position,
    );

    return DatabaseEventCreateField(payload).send();
  }

  Future<Either<TypeOptionPB, FlowyError>> insertBefore({
    FieldType fieldType = FieldType.RichText,
    String? fieldName,
    Uint8List? typeOptionData,
  }) {
    return createField(
      viewId: viewId,
      fieldType: fieldType,
      fieldName: fieldName,
      typeOptionData: typeOptionData,
      position: OrderObjectPositionPB(
        position: OrderObjectPositionTypePB.Before,
        objectId: fieldId,
      ),
    );
  }

  Future<Either<TypeOptionPB, FlowyError>> insertAfter({
    FieldType fieldType = FieldType.RichText,
    String? fieldName,
    Uint8List? typeOptionData,
  }) {
    return createField(
      viewId: viewId,
      fieldType: fieldType,
      fieldName: fieldName,
      typeOptionData: typeOptionData,
      position: OrderObjectPositionPB(
        position: OrderObjectPositionTypePB.Before,
        objectId: fieldId,
      ),
    );
  }

  static Future<Either<Unit, FlowyError>> moveField({
    required String viewId,
    required String fromFieldId,
    required String toFieldId,
  }) {
    final payload = MoveFieldPayloadPB(
      viewId: viewId,
      fromFieldId: fromFieldId,
      toFieldId: toFieldId,
    );

    return DatabaseEventMoveField(payload).send();
  }

  static Future<Either<Unit, FlowyError>> deleteField({
    required String viewId,
    required String fieldId,
  }) {
    final payload = DeleteFieldPayloadPB(
      viewId: viewId,
      fieldId: fieldId,
    );

    return DatabaseEventDeleteField(payload).send();
  }

  static Future<Either<Unit, FlowyError>> duplicateField({
    required String viewId,
    required String fieldId,
  }) {
    final payload = DuplicateFieldPayloadPB(
      viewId: viewId,
      fieldId: fieldId,
    );

    return DatabaseEventDuplicateField(payload).send();
  }

  Future<Either<Unit, FlowyError>> updateField({
    String? name,
    bool? frozen,
  }) {
    final payload = FieldChangesetPB.create()
      ..viewId = viewId
      ..fieldId = fieldId;

    if (name != null) {
      payload.name = name;
    }

    if (frozen != null) {
      payload.frozen = frozen;
    }

    return DatabaseEventUpdateField(payload).send();
  }

  static Future<Either<Unit, FlowyError>> updateFieldTypeOption({
    required String viewId,
    required String fieldId,
    required List<int> typeOptionData,
  }) {
    final payload = TypeOptionChangesetPB.create()
      ..viewId = viewId
      ..fieldId = fieldId
      ..typeOptionData = typeOptionData;

    return DatabaseEventUpdateFieldTypeOption(payload).send();
  }

  Future<Either<Unit, FlowyError>> updateFieldType({
    required FieldType fieldType,
  }) {
    final payload = UpdateFieldTypePayloadPB.create()
      ..viewId = viewId
      ..fieldId = fieldId
      ..fieldType = fieldType;

    return DatabaseEventUpdateFieldType(payload).send();
  }

  Future<Either<Unit, FlowyError>> delete() {
    return deleteField(viewId: viewId, fieldId: fieldId);
  }

  Future<Either<Unit, FlowyError>> duplicate() {
    return duplicateField(viewId: viewId, fieldId: fieldId);
  }

  Future<Either<TypeOptionPB, FlowyError>> getFieldTypeOptionData({
    required FieldType fieldType,
  }) {
    final payload = TypeOptionPathPB.create()
      ..viewId = viewId
      ..fieldId = fieldId
      ..fieldType = fieldType;
    return DatabaseEventGetTypeOption(payload).send().then((result) {
      return result.fold(
        (data) => left(data),
        (err) => right(err),
      );
    });
  }

  /// Returns the primary field of the view.
  static Future<Either<FieldPB, FlowyError>> getPrimaryField({
    required String viewId,
  }) {
    final payload = DatabaseViewIdPB.create()..value = viewId;
    return DatabaseEventGetPrimaryField(payload).send();
  }
}

File: frontend/appflowy_tauri/src/appflowy_app/components/database/application/field/field_service.ts
import {
  CreateFieldPayloadPB,
  DeleteFieldPayloadPB,
  DuplicateFieldPayloadPB,
  FieldChangesetPB,
  FieldType,
  GetFieldPayloadPB,
  MoveFieldPayloadPB,
  RepeatedFieldIdPB,
  UpdateFieldTypePayloadPB,
  FieldSettingsChangesetPB,
  FieldVisibility,
  DatabaseViewIdPB,
  CreateFieldPosition,
} from '@/services/backend';
import {
  DatabaseEventDuplicateField,
  DatabaseEventUpdateField,
  DatabaseEventUpdateFieldType,
  DatabaseEventMoveField,
  DatabaseEventGetFields,
  DatabaseEventDeleteField,
  DatabaseEventCreateField,
  DatabaseEventUpdateFieldSettings,
  DatabaseEventGetAllFieldSettings,
} from '@/services/backend/events/flowy-database2';
import { Field, pbToField } from './field_types';
import { getTypeOption } from './type_option';
import { Database } from '$app/components/database/application';

export async function getFields(
  viewId: string,
  fieldIds?: string[]
): Promise<{
  fields: Field[];
  typeOptions: Database['typeOptions'];
}> {
  const payload = GetFieldPayloadPB.fromObject({
    view_id: viewId,
    field_ids: fieldIds
      ? RepeatedFieldIdPB.fromObject({
          items: fieldIds.map((fieldId) => ({ field_id: fieldId })),
        })
      : undefined,
  });

  const result = await DatabaseEventGetFields(payload);

  const getSettingsPayload = DatabaseViewIdPB.fromObject({
    value: viewId,
  });

  const settings = await DatabaseEventGetAllFieldSettings(getSettingsPayload);

  if (settings.ok === false || result.ok === false) {
    return Promise.reject('Failed to get fields');
  }

  const typeOptions: Database['typeOptions'] = {};

  const fields = await Promise.all(
    result.val.items.map(async (item) => {
      const setting = settings.val.items.find((setting) => setting.field_id === item.id);

      const field = pbToField(item);

      const typeOption = await getTypeOption(viewId, item.id, item.field_type);

      if (typeOption) {
        typeOptions[item.id] = typeOption;
      }

      return {
        ...field,
        visibility: setting?.visibility,
        width: setting?.width,
      };
    })
  );

  return { fields, typeOptions };
}

export async function createField({
  viewId,
  targetFieldId,
  fieldPosition,
  fieldType,
  data,
}: {
  viewId: string;
  targetFieldId?: string;
  fieldPosition?: CreateFieldPosition;
  fieldType?: FieldType;
  data?: Uint8Array;
}): Promise<Field> {
  const payload = CreateFieldPayloadPB.fromObject({
    view_id: viewId,
    field_type: fieldType,
    type_option_data: data,
    target_field_id: targetFieldId,
    field_position: fieldPosition,
  });

  const result = await DatabaseEventCreateField(payload);

  if (result.ok === false) {
    return Promise.reject('Failed to create field');
  }

  return pbToField(result.val.field);
}

export async function duplicateField(viewId: string, fieldId: string): Promise<void> {
  const payload = DuplicateFieldPayloadPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
  });

  const result = await DatabaseEventDuplicateField(payload);

  if (result.ok === false) {
    return Promise.reject('Failed to duplicate field');
  }

  return result.val;
}

export async function updateField(
  viewId: string,
  fieldId: string,
  data: {
    name?: string;
    desc?: string;
  }
): Promise<void> {
  const payload = FieldChangesetPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
    ...data,
  });

  const result = await DatabaseEventUpdateField(payload);

  return result.unwrap();
}

export async function updateFieldType(viewId: string, fieldId: string, fieldType: FieldType): Promise<void> {
  const payload = UpdateFieldTypePayloadPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
    field_type: fieldType,
  });

  const result = await DatabaseEventUpdateFieldType(payload);

  return result.unwrap();
}

export async function moveField(viewId: string, fieldId: string, fromIndex: number, toIndex: number): Promise<void> {
  const payload = MoveFieldPayloadPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
    from_index: fromIndex,
    to_index: toIndex,
  });

  const result = await DatabaseEventMoveField(payload);

  return result.unwrap();
}

export async function deleteField(viewId: string, fieldId: string): Promise<void> {
  const payload = DeleteFieldPayloadPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
  });

  const result = await DatabaseEventDeleteField(payload);

  return result.unwrap();
}

export async function updateFieldSetting(
  viewId: string,
  fieldId: string,
  settings: {
    visibility?: FieldVisibility;
    width?: number;
  }
): Promise<void> {
  const payload = FieldSettingsChangesetPB.fromObject({
    view_id: viewId,
    field_id: fieldId,
    ...settings,
  });

  const result = await DatabaseEventUpdateFieldSettings(payload);

  if (result.ok === false) {
    return Promise.reject('Failed to update field settings');
  }

  return result.val;
}

export const reorderFields = (list: Field[], startIndex: number, endIndex: number) => {
  const result = Array.from(list);
  const [removed] = result.splice(startIndex, 1);

  result.splice(endIndex, 0, removed);

  return result;
};
